### PR TITLE
CORE-6619: move composite key implementation out of API

### PR DESCRIPTION
**Pull request by dickon**
CORE-6619: move composite key implementation out of API

This is a step toward removing Bouncy Castle from our public
API in corda-api, and also gets some of the implementation detail
(and quite a few lines of code) out of the public API.

### CODE REVIEW

File: crypto/src/main/kotlin/net/corda/v5/crypto/CompositeKeyGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package net.corda.v5.crypto
+
+import java.security.PublicKey
+
+interface CompositeKeyGenerator {
```
Comment (alexey-kadyrov-r3) [inline crypto/src/main/kotlin/net/corda/v5/crypto/CompositeKeyGenerator.kt]: As you have `CompositeKeyBuilder` this interface (CompositeKeyGenerator) can be moved to the runtime-os (to crypto-core)

File: crypto/src/main/kotlin/net/corda/v5/crypto/NodeAndWeight.kt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package net.corda.v5.crypto
+
+import java.security.PublicKey
+
+data class NodeAndWeight(val node: PublicKey, val weight: Int = 1)
```
Comment (alexey-kadyrov-r3) [inline crypto/src/main/kotlin/net/corda/v5/crypto/NodeAndWeight.kt]: Name it `CompositeKeyNodeAndWeight` as when it's standalone class the name as now doesn't convey too much

Comment (alexey-kadyrov-r3) [reply crypto/src/main/kotlin/net/corda/v5/crypto/NodeAndWeight.kt:5]: And I think that class can be moved to runtme-os (to crypto-core or maybe even crypto-impl) as it's not being used by any publicly exposed classes/interface - it became internal implementation details

File: application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyNodeAndWeight.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package net.corda.v5.application.crypto
+
+import java.security.PublicKey
+
+data class CompositeKeyNodeAndWeight(val node: PublicKey, val weight: Int = 1) {
```
Comment (alexey-kadyrov-r3) [inline application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyNodeAndWeight.kt]: As a DTO that class can be moved to corda-crypto

File: crypto/src/main/kotlin/net/corda/v5/crypto/NodeAndWeight.kt (hunk 2 above)
Comment (dickon) [reply crypto/src/main/kotlin/net/corda/v5/crypto/NodeAndWeight.kt]: done

File: crypto/src/main/kotlin/net/corda/v5/crypto/CompositeKeyGenerator.kt (hunk 1 above)
Comment (dickon) [reply crypto/src/main/kotlin/net/corda/v5/crypto/CompositeKeyGenerator.kt]: Resovled, via further design iteration.

File: application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyNodeAndWeight.kt (hunk 3 above)
Comment (dickon) [reply application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyNodeAndWeight.kt]: done

File: application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package net.corda.v5.application.crypto
+
+import net.corda.v5.crypto.CompositeKeyNodeAndWeight
+import java.security.PublicKey
+
+interface CompositeKeyGenerator {
```
Comment (alexey-kadyrov-r3) [inline application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyGenerator.kt]: Add the interface level KDocs

File: crypto/src/main/kotlin/net/corda/v5/crypto/CompositeKeyNodeAndWeight.kt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package net.corda.v5.crypto
+
+import java.security.PublicKey
+
+data class CompositeKeyNodeAndWeight(val node: PublicKey, val weight: Int = 1) {
```
Comment (alexey-kadyrov-r3) [inline crypto/src/main/kotlin/net/corda/v5/crypto/CompositeKeyNodeAndWeight.kt]: Can you add the overload constructor for the default weight

Comment (dickon) [reply crypto/src/main/kotlin/net/corda/v5/crypto/CompositeKeyNodeAndWeight.kt]: done

File: application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyGenerator.kt (hunk 4 above)
Comment (dickon) [reply application/src/main/kotlin/net/corda/v5/application/crypto/CompositeKeyGenerator.kt]: Done

**Review by alexey-kadyrov-r3 — DISMISSED** [review]
lgtm

**Review by alexey-kadyrov-r3 — APPROVED** [review]
LGTM

**Review by mcgovc — APPROVED** [review]
LGTM - (Gradle changes)

**Pull request merged**